src/main/java/org/javarosa/core/model/actions/setlocation/SetLocationAction.java

```diff
@@ -0,0 +1,120 @@
+import org.javarosa.core.model.data.AnswerDataFactory;
+import org.javarosa.core.model.data.IAnswerData;
+import org.javarosa.core.model.instance.AbstractTreeElement;
+import org.javarosa.core.model.instance.TreeReference;
+import org.javarosa.core.util.externalizable.DeserializationException;
+import org.javarosa.core.util.externalizable.ExtUtil;
+import org.javarosa.core.util.externalizable.ExtWrapNullable;
+import org.javarosa.core.util.externalizable.PrototypeFactory;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+
+/**
+ * Abstract implementation of odk:setlocation action. Concrete implementations must:
+ * - provide a way to request location updates when the action is triggered
+ * - use {@link #saveLocationValue(String)} to write a location to the model
+ * - provide a no-argument constructor with no body for serialization
+ * - get registered by the {@link org.javarosa.core.services.PrototypeManager}
+ */
+public abstract class SetLocationAction extends Action {
+    private TreeReference targetReference;
+    private TreeReference contextualizedTargetReference;
+
+    private FormDef formDef;
+
+    public SetLocationAction() {
+        // empty body for serialization
+    }
+
+    public SetLocationAction(TreeReference targetReference) {
+        super(SetLocationActionHandler.ELEMENT_NAME);
+        setTargetReference(targetReference);
+    }
+
+    public TreeReference getTargetReference() {
+        return targetReference;
+    }
+
+    public TreeReference getContextualizedTargetReference() {
+        return contextualizedTargetReference;
+    }
+
+    public void setTargetReference(TreeReference targetReference) {
+        this.targetReference = targetReference;
+    }
+
+    @Override
+    public final TreeReference processAction(FormDef model, TreeReference contextRef) {
+        this.formDef = model;
+        contextualizedTargetReference = contextRef == null ? this.targetReference
+            : this.targetReference.contextualize(contextRef);
+
+        requestLocationUpdates();
+
+        return contextualizedTargetReference;
+    }
+
+    /**
+     * Client-specific location request. Implementations could immediately read a location and write it to the model if
+     * available or could initiate a request and asynchronously write the location when available.
+     */
+    public abstract void requestLocationUpdates();
+
+    /**
+     * Save the location to the model.
+     */
+    public final void saveLocationValue(String location) {
+        EvaluationContext context = new EvaluationContext(formDef.getEvaluationContext(), contextualizedTargetReference);
+        AbstractTreeElement node = context.resolveReference(contextualizedTargetReference);
+        if (node != null) {
+            int dataType = node.getDataType();
+            IAnswerData val = Recalculate.wrapData(location, dataType);
+            if (val == null) {
```

Comment:
To avoid conditional logic, you could do this instead:
```java
IAnswerData answer = val != null
    ? AnswerDataFactory.templateByDataType(dataType).cast(val.uncast())
    : null;
formDef.setValue(answer, contextualizedTargetReference, true);
```

This way we explicitly separate the moving part (conditionally obtaining the answer) from the fixed part (setting the value in the form def)

Reply:
You and @dcbriccetti really love your ternary! 😊 I see the appeal in this case. Done.

Reply:
Though I think the ? belongs on the same line as the conditional expression, as in English. But if everybody were like me the world would be great^h^h^h^h^hboring.
:-)

Reply:
😂Nice to see some `^H^H^H`.

Reply:
Today I learned something new :) `^h`